File: games/shifter/car.py
"""
Physique voiture – portage fidèle du Car.js du POC Phaser.
"""
from config import (
    AIR_RES, ROLL_RES,
    SHIFT_PERF, SHIFT_GOOD, SHIFT_MULT, SHIFT_BONUS_DUR,
    RACE_DISTANCE,
)


class Car:
    def __init__(self, data: dict):
        s = data["stats"]
        self.data         = data
        self.name         = data["name"]
        self.color        = data["col"]
        self.max_power    = s["power"]
        self.weight       = s["weight"]
        self.max_gears    = s["gears"]
        self.max_rpm      = s["maxRPM"]
        self.opt_rpm      = s["optRPM"]
        self.shift_time   = s["shiftT"]
        self.curve        = data["curve"]
        self.ratios       = data["ratios"]

        # État dynamique
        self.rpm          = 1000.0
        self.speed        = 0.0    # km/h
        self.gear         = 1
        self.position     = 0.0   # mètres
        self.race_time    = 0.0   # secondes

        self.is_shifting      = False
        self.shift_cooldown   = 0.0
        self.shift_bonus      = 1.0
        self.shift_bonus_timer = 0.0
        self.shift_quality    = None  # 'PERFECT' | 'GOOD' | 'BAD' | None

        self.has_started  = False
        self.has_finished = False
        self.best_speed   = 0.0

        # Stats de course
        self.perfect_shifts = 0
        self.good_shifts    = 0
        self.bad_shifts     = 0

    # ── Courbe de puissance ────────────────────────────────────────────────────
    def power_at_rpm(self, rpm: float) -> float:
        rpm = max(1000.0, min(self.max_rpm, rpm))
        c = self.curve
        for i in range(len(c) - 1):
            r1, p1 = c[i]
            r2, p2 = c[i + 1]
            if r1 <= rpm <= r2:
                t = (rpm - r1) / (r2 - r1)
                return p1 + (p2 - p1) * t
        return c[-1][1]

    def calc_rpm(self, speed: float, gear: int) -> float:
        if speed < 1:
            return 1000.0
        # Constante abaissée (25 au lieu de 40) : chaque vitesse couvre
        # une plage de speed plus large → moins de shifts à spammer
        rpm = speed * self.ratios[gear - 1] * 25.0
        return max(1000.0, min(rpm, self.max_rpm + 500))

    # ── Changements de vitesse ─────────────────────────────────────────────────
    def shift_up(self):
        if self.gear < self.max_gears and not self.is_shifting:
            self.is_shifting = True
            old_rpm = self.rpm
            self.gear += 1
            self.rpm *= 0.7
            self._eval_shift(old_rpm)
            self.shift_cooldown = self.shift_time

    def shift_down(self):
        if self.gear > 1 and not self.is_shifting:
            self.is_shifting = True
            self.gear -= 1
            self.rpm = min(self.rpm * 1.4, self.max_rpm)
            self.shift_cooldown = self.shift_time

    def _eval_shift(self, rpm_before: float):
        diff = abs(rpm_before - self.opt_rpm)
        if diff <= SHIFT_PERF:
            q = 'PERFECT'
            self.perfect_shifts += 1
        elif diff <= SHIFT_GOOD:
            q = 'GOOD'
            self.good_shifts += 1
        else:
            q = 'BAD'
            self.bad_shifts += 1
        self.shift_quality     = q
        self.shift_bonus       = SHIFT_MULT[q]
        self.shift_bonus_timer = SHIFT_BONUS_DUR

    # ── Mise à jour physique ───────────────────────────────────────────────────
    def update(self, dt: float):
        if not self.has_started or self.has_finished:
            return

        # Cooldown de changement de vitesse
        if self.is_shifting:
            self.shift_cooldown -= dt
            if self.shift_cooldown <= 0:
                self.is_shifting    = False
                self.shift_cooldown = 0.0

        # Timer du bonus de shift
        if self.shift_bonus_timer > 0:
            self.shift_bonus_timer -= dt
            if self.shift_bonus_timer <= 0:
                self.shift_bonus   = 1.0
                self.shift_quality = None

        # ── Physique (identique au Car.js) ────────────────────────────────────
        power_hp = self.power_at_rpm(self.rpm) * self.shift_bonus
        power_w  = power_hp * 745.7
        vel_ms   = self.speed / 3.6

        wheel_force = power_w / vel_ms if vel_ms > 0.1 else power_w * 10.0
        accel       = wheel_force / self.weight

        air_drag     = AIR_RES  * self.speed ** 2
        rolling_drag = ROLL_RES * self.weight * 9.81
        total_drag   = (air_drag + rolling_drag) / self.weight
        net_accel    = accel - total_drag

        if self.is_shifting:
            net_accel *= 0.3
        if self.rpm >= self.max_rpm:
            net_accel *= 0.2

        self.speed = max(0.0, self.speed + net_accel * dt * 3.6)
        self.rpm   = self.calc_rpm(self.speed, self.gear)

        self.position  += (self.speed / 3.6) * dt
        self.race_time += dt

        if self.speed > self.best_speed:
            self.best_speed = self.speed

        if self.position >= RACE_DISTANCE:
            self.position    = RACE_DISTANCE
            self.has_finished = True

    def start(self):
        self.has_started = True

    def reset(self):
        self.__init__(self.data)

File: poc/shifter/src/entities/Car.js
import { GAME_CONFIG } from '../config.js';

export class Car {
    constructor(carData) {
        this.id = carData.id;
        this.name = carData.name;
        this.category = carData.category;
        this.description = carData.description;
        this.sprite = carData.sprite;

        // Stats
        this.maxPower = carData.stats.power;
        this.weight = carData.stats.weight;
        this.maxGears = carData.stats.gears;
        this.maxRPM = carData.stats.maxRPM;
        this.optimalShiftRPM = carData.stats.optimalShiftRPM;
        this.shiftTime = carData.stats.shiftTime;

        // Power curve (array of [rpm, power] points)
        this.powerCurve = carData.powerCurve;
        this.gearRatios = carData.gearRatios;

        // Dynamic state
        this.rpm = 1000;
        this.speed = 0;           // km/h
        this.gear = 1;
        this.position = 0;        // meters
        this.time = 0;            // seconds
        this.isShifting = false;
        this.shiftBonus = 1.0;
        this.shiftBonusTimer = 0;
        this.hasStarted = false;
        this.hasFinished = false;
        this.earlyStartPenalty = false;
    }

    /**
     * Get power at specific RPM using linear interpolation
     */
    getPowerAtRPM(rpm) {
        // Clamp RPM
        rpm = Math.max(1000, Math.min(this.maxRPM, rpm));

        // Find the two points to interpolate between
        for (let i = 0; i < this.powerCurve.length - 1; i++) {
            const [rpm1, power1] = this.powerCurve[i];
            const [rpm2, power2] = this.powerCurve[i + 1];

            if (rpm >= rpm1 && rpm <= rpm2) {
                // Linear interpolation
                const t = (rpm - rpm1) / (rpm2 - rpm1);
                return power1 + (power2 - power1) * t;
            }
        }

        // If beyond last point, return last power value
        return this.powerCurve[this.powerCurve.length - 1][1];
    }

    /**
     * Calculate RPM based on speed and current gear
     */
    calculateRPM(speed, gear) {
        // Formule réaliste: RPM = (vitesse_m/s * ratio * finalDrive * 60) / (2 * pi * wheelRadius)
        // Simplifié: RPM = speed_kmh * gearRatio * constant
        if (speed < 1) return 1000; // Idle RPM

        const constant = 40; // Constante de tuning calibrée
        const rpm = speed * this.gearRatios[gear - 1] * constant;
        return Math.max(1000, Math.min(rpm, this.maxRPM + 500));
    }

    /**
     * Shift up one gear
     */
    shiftUp() {
        if (this.gear < this.maxGears && !this.isShifting) {
            this.isShifting = true;
            const oldGear = this.gear;
            this.gear++;

            // RPM drop when shifting up
            this.rpm = this.rpm * 0.7;

            // Calculate shift quality
            this.evaluateShift(oldGear);

            // Shift takes time
            setTimeout(() => {
                this.isShifting = false;
            }, this.shiftTime * 1000);
        }
    }

    /**
     * Shift down one gear
     */
    shiftDown() {
        if (this.gear > 1 && !this.isShifting) {
            this.isShifting = true;
            this.gear--;

            // RPM increase when shifting down (but capped at maxRPM)
            this.rpm = Math.min(this.rpm * 1.4, this.maxRPM);

            setTimeout(() => {
                this.isShifting = false;
            }, this.shiftTime * 1000);
        }
    }

    /**
     * Evaluate shift quality and apply bonuses/penalties
     */
    evaluateShift(gearBeforeShift) {
        const rpmDiff = Math.abs(this.rpm / 0.7 - this.optimalShiftRPM); // Reverse the drop
        const { perfect, good } = GAME_CONFIG.shiftZones;
        const { bonuses, duration } = GAME_CONFIG.shiftZones;

        if (rpmDiff <= perfect) {
            // Perfect shift!
            this.shiftBonus = bonuses.perfect;
            this.shiftQuality = 'PERFECT';
        } else if (rpmDiff <= good) {
            // Good shift
            this.shiftBonus = bonuses.good;
            this.shiftQuality = 'GOOD';
        } else {
            // Bad shift
            this.shiftBonus = bonuses.bad;
            this.shiftQuality = 'BAD';
        }

        // Reset bonus after duration
        this.shiftBonusTimer = duration;
    }

    /**
     * Update car physics
     */
    update(deltaTime) {
        if (!this.hasStarted || this.hasFinished) return;

        // Update shift bonus timer
        if (this.shiftBonusTimer > 0) {
            this.shiftBonusTimer -= deltaTime * 1000;
            if (this.shiftBonusTimer <= 0) {
                this.shiftBonus = 1.0;
                this.shiftQuality = null;
            }
        }

        // Get current power from curve (en chevaux)
        let powerHP = this.getPowerAtRPM(this.rpm) * this.shiftBonus;

        // Conversion power to force: Power(HP) -> Power(Watts) -> Force(N)
        // 1 HP = 745.7 Watts
        // Power(W) = Force(N) * Velocity(m/s)
        // Force = Power / Velocity
        const powerWatts = powerHP * 745.7;
        const velocityMS = this.speed / 3.6; // km/h to m/s

        // Force disponible (éviter division par zéro)
        const wheelForce = velocityMS > 0.1 ? powerWatts / velocityMS : powerWatts * 10;

        // Accélération: F = m * a => a = F / m
        let acceleration = wheelForce / this.weight;

        // Résistances (air + friction)
        const airDrag = GAME_CONFIG.airResistance * this.speed * this.speed;
        const rollingDrag = GAME_CONFIG.rollingResistance * this.weight * 9.81;
        const totalDrag = airDrag + rollingDrag;

        // Accélération nette
        let netAcceleration = acceleration - totalDrag / this.weight;

        // Pénalité pendant le changement de vitesse
        if (this.isShifting) {
            netAcceleration *= 0.3;
        }

        // Rev limiter (coupe puissance si on tape le limiteur)
        if (this.rpm >= this.maxRPM) {
            netAcceleration *= 0.2;
        }

        // Update speed (m/s²)
        this.speed += netAcceleration * deltaTime * 3.6; // Convert acceleration to km/h
        this.speed = Math.max(0, this.speed);

        // Update RPM based on speed and gear
        this.rpm = this.calculateRPM(this.speed, this.gear);

        // Update position
        this.position += (this.speed / 3.6) * deltaTime; // Convert km/h to m/s

        // Update time
        this.time += deltaTime;

        // Check if finished
        if (this.position >= GAME_CONFIG.raceDistance) {
            this.hasFinished = true;
            this.position = GAME_CONFIG.raceDistance;
        }
    }

    /**
     * Start the race
     */
    start() {
        this.hasStarted = true;
    }

    /**
     * Penalize for early start
     */
    penalizeEarlyStart() {
        this.earlyStartPenalty = true;
        this.time += GAME_CONFIG.startSequence.earlyPenalty / 1000;
    }

    /**
     * Reset car to initial state
     */
    reset() {
        this.rpm = 1000;
        this.speed = 0;
        this.gear = 1;
        this.position = 0;
        this.time = 0;
        this.isShifting = false;
        this.shiftBonus = 1.0;
        this.shiftBonusTimer = 0;
        this.hasStarted = false;
        this.hasFinished = false;
        this.earlyStartPenalty = false;
        this.shiftQuality = null;
    }
}
